### PR TITLE
Support for all K3s args

### PR DESCRIPTION
**Pull request by LarsBingBong**
So that an KCLI end-user can specify
K3s configuration & installation args as wanted. Not only --disabled

.gitignore:
- small brush-up
- add kubectl to be ignored. Useful when developing locally

config.py:
- conditional on the k3s worker join.sh script. In order to support both cmds in vmrules in a k3s deploy plan file
- small brush-ups
- the ability to send in the name of the cluster, to reach join.sh

__init__.py k3s:
- supports specifying the version for workers
- only generate the join.sh file if any workers are to be installed
- k3s install & config args for workers
- remove cURL as it's not an requirement for K3s & is a security liability

bootstrap.sh:
- remove cURL, see reason above
- support all k3s conf. & install args

kcli_default:
- alphabetic order of params
- extra info
- added k3s args parameters
-- k3s_extra_args_all
-- k3s_extra_args_masters
-- k3s_extra_args_workers

masters.sh:
- support all k3s install & conf. args
- remove cURL, see reason above

workers.yml:
- use the scripts clause approach in order to support cmds in a k3s plan

This feature also:
- removes the sdn parameter
 KCLI users can specify to remove
 the default CNI == flannel with k3s install
 & conf. args & then after the fact install
 the CNI they want

### CODE REVIEW

File: kvirt/config.py [hunk 1]
```diff
@@ -300,7 +300,8 @@ def __init__(self, client=None, debug=False, quiet=False, region=None, zone=None
         self.overrides.update(config_data)
 
     def create_vm(self, name, profile, overrides={}, customprofile={}, k=None,
-                  plan='kvirt', basedir='.', client=None, onfly=None, wait=False, onlyassets=False):
+                  plan='kvirt', basedir='.', client=None, onfly=None, wait=False,
+                  onlyassets=False, cluster=None):
```
Comment (karmab) [inline kvirt/config.py:304]: why introduce this variable? 
it seems to be solely for the purpose of changing k3s workflow, in which case you can instead rely on the fact that overrides dict will contain a kube variable (pretty much your cluster variable) and a kubetype (k3s, generic or openshift)

File: kvirt/config.py [hunk 2]
```diff
@@ -655,13 +656,23 @@ def create_vm(self, name, profile, overrides={}, customprofile={}, k=None,
                         os.makedirs(destdir, exist_ok=True)
                     common.fetch("%s/%s" % (onfly, script), destdir)
                 script = os.path.expanduser(script)
-                if basedir != '.':
+                if basedir != '.' and not script.endswith('join.sh'):
```
Comment (karmab) [inline kvirt/config.py:659]: i think it s wrong to handle a script named join.sh in a different way

File: kvirt/config.py [hunk 3]
```diff
@@ -655,13 +656,23 @@ def create_vm(self, name, profile, overrides={}, customprofile={}, k=None,
...
                     script = '%s/%s' % (basedir, script)
+                else:
```
Comment (karmab) [inline kvirt/config.py:661]: modifications related to k3s should only go in the k3s directory

File: kvirt/config.py [hunk 4]
```diff
@@ -655,13 +656,23 @@ def create_vm(self, name, profile, overrides={}, customprofile={}, k=None,
...
+                    #if script.endswith('join.sh'):
+                    script = os.path.expanduser("~/.kcli/clusters/%s/%s" % (cluster, scriptname))
+                
+                # Final conditionals before templating the lines of the script into cloud-init runcmd 
                 if script.endswith('register.sh') and skip_rhnregister_script:
                     continue
                 elif not os.path.exists(script):
                     return {'result': 'failure', 'reason': "Script %s not found" % script}
                 else:
+                    # Templating the scriptlines into cloud-init runcmd values
```
Comment (karmab) [inline kvirt/config.py:675]: you added the comment just to clarify the purpose of this section?

File: kvirt/config.py [hunk 5]
```diff
@@ -1264,7 +1275,7 @@ def revert_plan(self, plan, snapshotname=None):
 
     def plan(self, plan, ansible=False, url=None, path=None, container=False, inputfile=None, inputstring=None,
              overrides={}, info=False, update=False, embedded=False, download=False, quiet=False, doc=False,
-             onlyassets=False, pre=True):
+             onlyassets=False, pre=True, cluster=None):
```
Comment (karmab) [inline kvirt/config.py:1278]: same comment as earlier, an additional cluster variables isnt needed

File: kvirt/config.py [hunk 6]
```diff
@@ -1556,7 +1567,7 @@ def plan(self, plan, ansible=False, url=None, path=None, container=False, inputf
                     continue
         if vmentries:
             if not onlyassets:
-                pprint("Deploying Vms...")
+                pprint("Deploying VM's...")
```
Comment (karmab) [inline kvirt/config.py:1570]: mind moving this to a dedicated PR, since this one is more tricky?

File: .gitignore [hunk 7]
```diff
@@ -1,18 +1,22 @@
+# Folder to ignore
```
Comment (karmab) [inline .gitignore:1]: mind moving this to a dedicated PR, since this one is more tricky?

File: kvirt/k3s/__init__.py [hunk 8]
```diff
@@ -74,8 +75,14 @@ def create(config, plandir, cluster, overrides):
                 os._exit(1)
     install_k3s_channel = data.get('install_k3s_channel', 'stable')
     if install_k3s_channel not in ['stable', 'latest', 'testing']:
-        error("Invalid K3S install channel %s" % install_k3s_channel)
+        error("Invalid K3s install channel %s" % install_k3s_channel)
```
Comment (karmab) [inline kvirt/k3s/__init__.py:78]: mind moving this to a dedicated PR, since this one is more tricky?

File: kvirt/k3s/__init__.py [hunk 9]
```diff
@@ -105,25 +112,39 @@ def create(config, plandir, cluster, overrides):
...
-    source, destination = "/root/kubeconfig", "%s/auth/kubeconfig" % clusterdir
-    scpcmd = scp(firstmaster, ip=firstmasterip, user='root', source=source, destination=destination,
-                 tunnel=config.tunnel, tunnelhost=config.tunnelhost, tunnelport=config.tunnelport,
-                 tunneluser=config.tunneluser, download=True, insecure=True, vmport=firstmastervmport)
-    os.system(scpcmd)
+    ################
+    # WORKER SETUP #
+    ################
     workers = data.get('workers', 0)
     if workers > 0:
```
Comment (karmab) [inline kvirt/k3s/__init__.py:119]: the join.sh script needs to be generated even when no workers are specified initially, to allow to scale after the fact

File: kvirt/k3s/__init__.py [hunk 10]
```diff
@@ -105,25 +112,39 @@ def create(config, plandir, cluster, overrides):
...
+            f.write("%s \n" % joincmd)
+            f.write("apt-get -y remove curl \n")
+        
+        # Deploy the workers
         pprint("Deploying workers")
         if 'name' in data:
             del data['name']
         os.chdir(os.path.expanduser("~/.kcli"))
-        config.plan(plan, inputfile='%s/workers.yml' % plandir, overrides=data)
+        config.plan(plan, inputfile='%s/workers.yml' % plandir, overrides=data, cluster=cluster)
```
Comment (karmab) [inline kvirt/k3s/__init__.py:138]: as commented before, in this case, data contains data['kube'], which is your cluster variable

File: kvirt/k3s/bootstrap.sh [hunk 11]
```diff
@@ -1,23 +1,22 @@
...
+curl -sfL https://get.k3s.io | INSTALL_K3S_CHANNEL={{ install_k3s_channel }} INSTALL_K3S_VERSION={{ install_k3s_version if install_k3s_version != "latest" else '' }} K3S_TOKEN={{ token }} sh -s - server --cluster-init {{ extra_args|join(" ") }}
 export IP={{ api_ip }}
 {% else %}
-curl -sfL https://get.k3s.io | {{ "INSTALL_K3S_EXEC='--flannel-backend=none'" if sdn != "flannel" else '' }} INSTALL_K3S_CHANNEL={{ install_k3s_channel }} INSTALL_K3S_VERSION={{ install_k3s_version if install_k3s_version != "latest" else '' }} K3S_TOKEN={{ token }} sh -s - server {{ extra_args|join(" ") }}
+curl -sfL https://get.k3s.io | INSTALL_K3S_CHANNEL={{ install_k3s_channel }} INSTALL_K3S_VERSION={{ install_k3s_version if install_k3s_version != "latest" else '' }} K3S_TOKEN={{ token }} sh -s - server {{ extra_args|join(" ") }}
 export IP=$(hostname -I | cut -f1 -d" ")
 {% endif %}
+apt-get -y remove curl
```
Comment (karmab) [inline kvirt/k3s/bootstrap.sh:16]: i find it misleading that one has to install curl as per k3s doc only to perform the install.
What if you want to troubleshoot and you see curl is not there? that would be quite misleading

File: kvirt/k3s/bootstrap.sh [hunk 12]
```diff
@@ -1,23 +1,22 @@
...
 export K3S_TOKEN=$(cat /var/lib/rancher/k3s/server/node-token)
 sed "s/127.0.0.1/$IP/" /etc/rancher/k3s/k3s.yaml > /root/kubeconfig
 if [ -d /root/manifests ] ; then
  mkdir -p /var/lib/rancher/k3s/server
  mv /root/manifests /var/lib/rancher/k3s/server
 fi
-{% if sdn == 'cilium' %}
```
Comment (karmab) [inline kvirt/k3s/bootstrap.sh:19]: alternatively, we could fix whatever is missing in cilium side instead of removing it. 
But if you think it doesnt make sense to keep it as is right now, let's instead open a PR to study its removal

File: kvirt/k3s/kcli_default.yml [hunk 13]
```diff
@@ -1,48 +1,51 @@
 info: |
-    Deploys the k3s kubernetes distrobution using an arbitrary number of masters and workers.
```
Comment (karmab) [inline kvirt/k3s/kcli_default.yml:2]: you just reorder the variables here? cos i have a hard time figuring out what was added based on the github diff

File: kvirt/k3s/workers.yml [hunk 14]
```diff
@@ -20,15 +20,8 @@
...
-  - path: /root/join.sh
-    origin: {{ basedir }}/clusters/{{ cluster }}/join.sh
- cmds:
- - apt-get -y install curl
-{% if sdn == 'cilium' %}
- - mount bpffs -t bpf /sys/fs/bpf
-{% endif %}
- - bash /root/join.sh
+ scripts:
+  - join.sh
```
Comment (karmab) [inline kvirt/k3s/workers.yml:24]: the path should be kept the same, otherwise this wont work when scaling (and it wouldnt on a new install even, i suspect you have it in your local directory which is what makes it work for you)

**Review by karmab — CHANGES_REQUESTED** [review]

File: kvirt/config.py (hunk 2 above)
Comment (larssb) [reply kvirt/config.py:659]: Yeah I don't love this approach either. But, there's already an exception for the register.sh script ... so I figured it wouldn't be that bad 😉

File: kvirt/config.py (hunk 3 above)
Comment (larssb) [reply kvirt/config.py:661]: Agreed .. just wasn't sure how-to implement that.

File: kvirt/config.py (hunk 4 above)
Comment (larssb) [reply kvirt/config.py:675]: Indeed. I felt that some comments would make wonders in that section. That's all.

File: kvirt/config.py (hunk 5 above)
Comment (larssb) [reply kvirt/config.py:1278]: Right ... had no idea into how `overrides` work. Still don't quite get how I would then "reach" the cluster name via the `overrides` var

File: kvirt/k3s/__init__.py (hunk 10 above)
Comment (larssb) [reply kvirt/k3s/__init__.py:138]: aaaah now I'm starting to get it 👍

File: kvirt/k3s/bootstrap.sh (hunk 11 above)
Comment (larssb) [reply kvirt/k3s/bootstrap.sh:16]: I agree. cURL is not a requirement for K3s itself - but for the installation of it. In the way we do it - the script method. But, still - cURL is not something I want on my K3s nodes, in regards to security.
One could just install cURL again.
I guess it's a matter of how opinionated we allow ourselves to be.

File: kvirt/k3s/bootstrap.sh (hunk 12 above)
Comment (larssb) [reply kvirt/k3s/bootstrap.sh:19]: Yeah agree to the part of your comment that points to the fact that this change shouldn't really be a part of this PR.

In regards to Cilium. It was very much my wish to look into using Cilium instead of flannel. For other users ... with so many choices for CNI's. I think it makes more sense to say:

1. you can leave the CNI as is - meaning what K3s comes with per default
2. or you can uninstall flannel with the K3s install and conf. args and then install whatever CNI you want

At the time, where I wished to get Cilium, I was still trying to grasp how much KCLI should automate for end-users. I guess that is still very much a subject up for debate.

File: kvirt/k3s/kcli_default.yml (hunk 13 above)
Comment (larssb) [reply kvirt/k3s/kcli_default.yml:2]: Yeah ... I thought that this little brush-up could make you a little agitated. All I'm trying to-do is make the file more readable. And finding 'x' variable when the file is in lexicographical order just finds me easier.
And I added the three K3s arg. array variables.

File: kvirt/k3s/workers.yml (hunk 14 above)
Comment (larssb) [reply kvirt/k3s/workers.yml:24]: nopes ... I don't have it in my local dir. It works because I in the config.py folder look at the join.sh file and update the path to the file - so that it points to where it's being generated, which is underneath the specific cluster folder in e.g. `~/.kcli/clusters/karmabs-cluster/join.sh` - so it should scale.

**Comment by larssb** [comment]
Loads of comments from me .. I'll wait for you trying to implement this yourself. And then we'll continue the good discussion.

File: kvirt/config.py (hunk 2 above)
Comment (karmab) [reply kvirt/config.py:659]: the exception for the register.sh script is to avoid duplicates from days where i didnt support natively rhnregister stuff, but i think it s actually time to remove it :)

**Comment by karmab** [comment]
handled in https://github.com/karmab/kcli/pull/286 rather with all your good feedback there!
